Move watch latency metric into drand client

petar opened the pull request.

willscott commented on client/metric.go:
```diff
@@ -0,0 +1,74 @@
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func newMetricController(chainInfo *chain.Info, b PrometheusBridge) *metricController {
```
more of this, ideally, can live in the `/metrics` package.

Can we have all the gauge / counters / etc. living there, and have a client gatherer method, parallel to how that package already provides a subset of metrics relevant to the group.

generic push code could also live in that package.

willscott commented on client/client.go:
```diff
@@ -213,3 +235,25 @@ func WithWatcher(wc WatcherCtor) Option {
 	}
 }
+
+// PrometheusBridge abstracts the Prometheus metric registration and push functionalities.
+type PrometheusBridge interface {
+	Register(prometheus.Collector) error
+	Push() error
+}
+
+// WithPrometheus specifies a prometheus system to be used for metric collection.
+func WithPrometheus(b PrometheusBridge) Option {
+	return func(cfg *clientConfig) error {
+		cfg.prometheus = b
+		return nil
+	}
+}
```
can this be simplified to `WithRegistry(prometheus.Registerer)`
https://github.com/prometheus/client_golang/blob/master/prometheus/registry.go#L97

willscott commented on client/client.go:
```diff
@@ -84,7 +86,22 @@ func makeClient(cfg clientConfig) (Client, error) {
+	c = newWatchAggregator(c, cfg.log)
+
+	if cfg.prometheus != nil {
+		if cfg.id == "" {
```
i would expect the registry passed in would use
use [`WrapRegistererWith`](https://godoc.org/github.com/prometheus/client_golang/prometheus#WrapRegistererWith) rather than this needing to be in the interface

willscott commented on client/metric.go:
```diff
@@ -0,0 +1,64 @@
+	c := &watchLatencyMetricClient{
+		Client: base,
+		ctl:    ctl,
+		watchLatency: prometheus.NewGauge(prometheus.GaugeOpts{
```
Probably want a single [`GaugeVec`](https://godoc.org/github.com/prometheus/client_golang/prometheus#NewGaugeVec) that specifies a label on the id of the specific watcher

willscott commented on cmd/demo-client/main.go:
```diff
@@ -159,3 +203,49 @@ func Watch(c *cli.Context, client client.Client) error {
+		go b.pushLoop()
+	}
+	if address != "" {
+		http.Handle("/metrics", promhttp.Handler())
```
should be an explicit collector. avoid the global one.
(The library will call `Register` and give you a collector. Currently this bridge stands in front of the implict global registry. You can make a `NewRegistery()`, that you then can ask for a [`promhttp.HandlerFor(registery)`](https://godoc.org/github.com/prometheus/client_golang/prometheus/promhttp#HandlerFor)
if you use the provided implementations you shouldn't have to re-implement shims for Register/mustregister/unregister

petar commented:
@willscott I've addressed all the comments. Take a look. Specifically, tell me if https://github.com/drand/drand/blob/petar/metric/client/metric.go#L55 looks correct (given the usage of WrapRegistererWith).

willscott commented on cmd/demo-client/main.go:
```diff
@@ -159,3 +206,53 @@ func Watch(c *cli.Context, client client.Client) error {
+}
+
+func (b *prometheusBridge) pushLoop() {
+	for {
```
have a way to cancel / life-time this?

petar replied:
Why? There is no use case for this.

willscott replied:
so that as we make tests for the client CLI they can cancel the context they've provided and have everything associated with it shut down

petar replied:
I think this is out of scope for this PR, because making the client be able to Shutdown requires a lot more than just fixing this pushLoop. None of the client components have shutdown logic yet. Fixing this should belong in a separate PR, I think.

willscott commented on cmd/demo-client/main.go:
```diff
@@ -159,3 +206,53 @@ func Watch(c *cli.Context, client client.Client) error {
 	return nil
 }
+
+func newPrometheusBridge(address string, gateway string, pushIntervalSec int64) prometheus.Registerer {
```
might be able to simplify this structure (this object takes a registry and exposes it either at an http location or by active pushing) by taking in a registry to expose.

petar replied:
Why is this necessary? This is a binary-private method whose only job is to create the registry out of command line arguments. What is the point of "generalizing" it to take a Registry, when I will have to then move the Registry creation code in some other function. There is only one use for this method, that's why it is private.

willscott replied:
I'm trying to push towards have a consistent pattern across the binaries.

the pattern on cli / relay are patterned to only expose rather than internally manage a registry. This one case is inverting that pattern, and I don't see why. It can instead directly take the `metrics.ClientMetrics` registry that follows the pattern (see #437), and directly either serve or push it. the duplicate structure managing a separate registry for the binary isn't needed.

petar replied:
This sounds reasonable, but it is also out of scope for this PR: The effort to make binaries similar (as well as to fold the client into the drand-cli command) are largish separate efforts. This PR is supposed to be tested in one hour in order to derisk it by verifying its functionality. I don't think we should keep adding goals to it in light of future objectives. (Go development is incremental after all, no?)

Also note what is going on here: This comment is justified by an entirely new PR (437) that was just introduced. This is a prime example of scope creep. I am not against fixing this, but PR's shouldn't keep on being extended longer and longer as they are being reviewed. This is really bad practice.

willscott commented on cmd/demo-client/main.go:
```diff
@@ -159,3 +206,53 @@ func Watch(c *cli.Context, client client.Client) error {
+	}
+}
+
+func (b *prometheusBridge) Register(x prometheus.Collector) error {
```
if the first member of the struct is the `Registry`, then the registry can directly implement these methods.
In the constructor, the pusher can use [`Gatherer`](https://godoc.org/github.com/prometheus/client_golang/prometheus/push#Pusher.Gatherer) to push whatever has been registered with the registry.

petar replied:
Done.
